From 02cd5c8ef49af5e30d113abfa44be3bd58ecc34c Mon Sep 17 00:00:00 2001
From: Alex Johnston <acjohnston@google.com>
Date: Thu, 6 Jan 2022 01:27:36 +0000
Subject: [PATCH] Add caller check to com.android.credentials.RESET

* Only the Car Settings app can reset credentials
  via com.android.credentials.RESET.
* com.android.credentials.INSTALL should still be
  callable by CertInstaller.

Manual testing steps:
* Install certificate via Settings
* Verify unable to reset certificates via test app
  provided in the bug (app-debug.apk)
* Verify able to reset certificates via Settings
* Verify com.android.credentials.INSTALL isn't changed

Bug: 200164168
Test: manual
Change-Id: Ibf6a02a35beed08c769fda9b3044d2e7933186dd
(cherry picked from commit f366a7f52cee0a50c1f672158c4a70d007b61cae)
(cherry picked from commit 6a6489935d203715a755b21b374e1e3b3085aa3f)
Merged-In:Ibf6a02a35beed08c769fda9b3044d2e7933186dd
---
 .../security/CredentialStorageActivity.java       | 15 ++++++++++++++-
 1 file changed, 14 insertions(+), 1 deletion(-)

diff --git a/src/com/android/car/settings/security/CredentialStorageActivity.java b/src/com/android/car/settings/security/CredentialStorageActivity.java
index 22e7fd7f0..f40fcccd3 100644
--- a/src/com/android/car/settings/security/CredentialStorageActivity.java
+++ b/src/com/android/car/settings/security/CredentialStorageActivity.java
@@ -76,7 +76,7 @@ public class CredentialStorageActivity extends FragmentActivity {
 
         Intent intent = getIntent();
         String action = intent.getAction();
-        if (ACTION_RESET.equals(action)) {
+        if (ACTION_RESET.equals(action) && checkCallerIsSelf()) {
             showResetConfirmationDialog();
         } else if (ACTION_INSTALL.equals(action) && checkCallerIsCertInstallerOrSelfInProfile()) {
             Bundle installBundle = intent.getExtras();
@@ -118,6 +118,19 @@ public class CredentialStorageActivity extends FragmentActivity {
         }
     }
 
+    /**
+     * Check that the caller is Settings.
+     */
+    private boolean checkCallerIsSelf() {
+        try {
+            return Process.myUid() == android.app.ActivityManager.getService()
+                    .getLaunchedFromUid(getActivityToken());
+        } catch (RemoteException re) {
+            // Error talking to ActivityManager, just give up
+            return false;
+        }
+    }
+
     /**
      * Check that the caller is either CertInstaller or Settings running in a profile of this user.
      */
-- 
2.17.1

